### PR TITLE
Add CI build with NVC++ compiling CUDA

### PR DESCRIPTION
**Pull request by dalg24**
Modify NVHPC build that currently used NVC++ as the host compiler to also use it to compiler CUDA

### CODE REVIEW

**Comment by crtrott** [comment]
@dalg24 can you rebase this so we don't have 3k line change?

**Comment by dalg24** [comment]
> @dalg24 can you rebase this so we don't have 3k line change?

I was planning on rebasing AFTER we make the CI past so we don't lose the references to the older builds.
Retest this please

File: core/unit_test/TestHostSharedPtrAccessOnDevice.hpp [hunk 1]
```diff
@@ -136,6 +136,12 @@ TEST(TEST_CATEGORY, host_shared_ptr_dereference_on_device) {
       static_cast<T*>(Kokkos::kokkos_malloc<MemorySpace>(sizeof(T))),
       [](T* p) { Kokkos::kokkos_free<MemorySpace>(p); });
 
+#if defined(KOKKOS_ENABLE_CUDA) && \
+    defined(KOKKOS_COMPILER_NVHPC)  // FIXME_NVHPC
+  if constexpr (std::is_same_v<TEST_EXECSPACE, Kokkos::Cuda>) {
+    GTEST_SKIP() << "FIXME please";
```
Comment (dalg24) [inline core/unit_test/TestHostSharedPtrAccessOnDevice.hpp:142]: This one I don't remember seeing on my local build

File: .jenkins [hunk 2]
```diff
@@ -66,22 +67,21 @@ pipeline {
...
                         OMP_PROC_BIND = 'spread'
-                        NVCC_WRAPPER_DEFAULT_COMPILER = 'nvc++'
                     }
                     steps {
                         sh '''rm -rf build && mkdir -p build && cd build && \
                               /opt/cmake/bin/cmake \
-                                -DCMAKE_BUILD_TYPE=Debug \
-                                -DCMAKE_CXX_COMPILER=$WORKSPACE/bin/nvcc_wrapper \
-                                -DCMAKE_CXX_FLAGS=-Werror \
+                                -DCMAKE_BUILD_TYPE=RelWithDebInfo \
```
Comment (dalg24) [inline .jenkins:74]: Doesn't build in `Debug`.  Has been reported.

File: .jenkins [hunk 3]
```diff
@@ -66,22 +67,21 @@ pipeline {
...
                                 -DCMAKE_CXX_STANDARD=17 \
+                                -DCMAKE_CXX_FLAGS="--diag_suppress=implicit_return_from_non_void_function" \
                                 -DKokkos_ARCH_NATIVE=ON \
                                 -DKokkos_ENABLE_COMPILER_WARNINGS=ON \
                                 -DKokkos_ENABLE_DEPRECATED_CODE_4=OFF \
                                 -DKokkos_ENABLE_TESTS=ON \
                                 -DKokkos_ENABLE_CUDA=ON \
                                 -DKokkos_ENABLE_CUDA_LAMBDA=ON \
-                                -DKokkos_ENABLE_OPENMP=ON \
+                                -DKokkos_ENABLE_OPENMP=OFF \
```
Comment (dalg24) [inline .jenkins:84]: Yields an error at link-time.  Has been reported and should be fixed in NVHPC 22.9

**Comment by dalg24** [comment]
Failure is unrelated (perf test on another build that is using NVCC)

File: algorithms/unit_tests/TestSort.hpp [hunk 4]
```diff
@@ -432,12 +432,20 @@ void test_issue_4978_sort() {
 template <class ExecutionSpace, typename KeyType>
 void test_sort(unsigned int N) {
   test_1D_sort<ExecutionSpace, KeyType>(N);
-  test_3D_sort<ExecutionSpace, KeyType>(N);
+#if defined(KOKKOS_ENABLE_CUDA) && \
+    defined(KOKKOS_COMPILER_NVHPC)  // FIXME_NVHPC
```
Comment (masterleinad) [inline algorithms/unit_tests/TestSort.hpp:436]: What's the problem? Do you just get wrong results?

File: algorithms/unit_tests/TestStdAlgorithmsAdjacentFind.cpp [hunk 5]
```diff
@@ -315,6 +315,12 @@ void run_all_scenarios() {
 }
 
 TEST(std_algorithms_nonmod_seq_ops, adjacent_find) {
+#if defined(KOKKOS_ENABLE_CUDA) && \
+    defined(KOKKOS_COMPILER_NVHPC)  // FIXME_NVHPC
+  if constexpr (std::is_same_v<exespace, Kokkos::Cuda>) {
+    GTEST_SKIP() << "FIXME please";
```
Comment (masterleinad) [inline algorithms/unit_tests/TestStdAlgorithmsAdjacentFind.cpp:321]: Wrong result? Compiler error? Please comment.

**Review by masterleinad — COMMENTED** [review]
IMHO, It would be valuable to elaborate on the FIXMEs some.

File: algorithms/unit_tests/TestStdAlgorithmsAdjacentFind.cpp (hunk 5 above)
Comment (dalg24) [reply algorithms/unit_tests/TestStdAlgorithmsAdjacentFind.cpp:321]: Wrong result, the `if constexpr` only guards the skip test, so it is being compiled

File: algorithms/unit_tests/TestSort.hpp (hunk 4 above)
Comment (dalg24) [reply algorithms/unit_tests/TestSort.hpp:436]: Wrong result

**Comment by dalg24** [comment]
> IMHO, It would be valuable to elaborate on the FIXMEs some.

What would you suggest we print?

File: algorithms/unit_tests/TestSort.hpp (hunk 4 above)
Comment (dalg24) [reply algorithms/unit_tests/TestSort.hpp:436]: And I see know the code is wrong I only meant to guard the runtime if

Comment (dalg24) [reply algorithms/unit_tests/TestSort.hpp:436]: > And I see know the code is wrong I only meant to guard the runtime if

Fixed

File: algorithms/unit_tests/TestStdAlgorithmsAdjacentFind.cpp (hunk 5 above)
Comment (dalg24) [reply algorithms/unit_tests/TestStdAlgorithmsAdjacentFind.cpp]: Updated

**Review by masterleinad — APPROVED** [review]

**Pull request merged**